### PR TITLE
Add mlflow logging

### PR DESCRIPTION
**Pull request by smithcommajoseph**
Adds support for logging model training data to an MLFlow instance.

Note: Alpha and Bravo models are well supported. Charlie and Alpha Ensemble models will require some work at time of writing. @haneslinger, I'd appreciate feedback re: how to add better logging for the above.

### CODE REVIEW

**Comment by haneslinger** [comment]
Could/should this replace `Training_history.csv` and `save_model`/`load_model`?

**Comment by stephen-frank** [comment]
How big is MLFlow, as in, how heavy of a dependency is it if we were to make it mandatory? Thinking about compactness of deploying Wattile via Docker. If it is not a heavy dependency, I don't necessarily have a problem with it being required if it makes everything simpler.

**Comment by smithcommajoseph** [comment]
TLDR: MLFlow support adds value, but making MLFlow mandatory adds complexity and could have adverse side effects. 

If we make MLFlow mandatory, Wattile will become more brittle and less tolerant of network outages. While the MLFLow SDK has a concept of resilient sessions, where it retries unresponsive endpoints n-times over a given period, it will throw errors if the server is not found or unresponsive. If this occurs when it is time to log model metrics or artifacts, then an operator may have to repeat the training, which has fiscal and temporal impacts (I'm all about saving time and money). Wattile's current ability to log artifacts to the file system without network dependency, except for the initial package downloads and data required to train the model(s), makes it resilient to intermittent network outages. Assuming all goes well during the training phase, this pattern ensures that Wattile presents an operator with a fully functional model and its artifacts. The implementation proposed in this PR attempts to log metrics and wholesale upload all artifacts to MLFlow for storage.

Additionally, MLFlow represents one more server that an organization will need to set up and maintain. Indeed, one can run MLFlow locally for smaller jobs, but in practice, folks will likely want to use Databricks, AWS, or in-house hardware to run this server, which feels like a burdensome dependency to add. The implementation proposed in this PR makes MLFlow 100% optional. It adds value and provides operators/users a way to track, rank, and store models, which is crucial to scalable ML workflows. But if folks want to bite only some of that off when they start experimenting with Wattile, then no worries, as all core functionality is available with or without MLFlow.

**Review by stephen-frank — APPROVED** [review]
I'm not going to pretend I understand all these changes enough to be a useful reviewer, but at least I didn't see any red flags!
